### PR TITLE
fix: Unable to submit booking form if Attendee Phone is used and then booking Fields are saved

### LINKED ISSUE #15727: [CAL-4025] Invalid Input received on Attendee Address
### Issue Summary

I'm currently helping a friend utilize Cal's in-line embed set up for an event. The event utilizes "In Person (Attendee Address)". When I test out the form after embedding it into the site it always throws an invalid input error on the Address section.

### Steps to Reproduce

1. Create Event for a calendar
2. Set location to "In Person (Attendee Address)"
3. Go to advanced tab
4. Click edit for Location inside Booking Questions
5. Change label for the question and save the event type
6. Go to event type link OR Embed into your site and test address

### Actual Results
Screenshot 2024-07-10 at 11 02 01 PM

Screenshot 2024-07-10 at 11 02 22 PM

Screenshot 2024-07-10 at 10 58 40 PM

### Expected Results

- The address be accepted

### Technical details

- Utilizing GoDaddy Website Builder
- Testing on Chrome latest version

### Evidence

- This was tested by embedding the in-line code given to me through cal.com to a Go Daddy Website Builder, then publishing the website and testing.

<sub>[CAL-4025](https://linear.app/calcom/issue/CAL-4025/invalid-input-received-on-attendee-address)</sub>

### PR DESCRIPTION
**Pull request by vijayraghav-io**
## What does this PR do?

- Fixes #15727 
- Fixes CAL-4025 

We now show a Badge in front of Location field to convey that it has come from the locations. The count is the number of locations used by the organizer for the event.
image

Also reverts #13989 but fixes the problem addressed by it in the correct way. The problems addressed:
- Ability to change the Location label.
- Ability to reorder the Location when there is just Attendee Address/Attendee Phone

Before Changes -
https://www.loom.com/share/dd1094dca81d48039de31521f4bc6561?sid=79e39a7e-fd3d-412c-a50a-ec85bf2444c4

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A - I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [x] N/A -  I confirm automated tests are in place that prove my fix is effective or that my feature works.

### CODE REVIEW

**Review by Amit91848 — CHANGES_REQUESTED** [review]
@vijayraghav-io this only prevents the error from being thrown. The core issue is still unresolved. It works fine until you add a label to the question, and this does nothing about it. We still need to figure out what actually is breaking it.

When you create a booking the where is no longer Attendee Address but changes to default.


https://github.com/calcom/cal.com/assets/74371312/28ed690d-9cb9-4083-81e3-9488da021795

**Comment by vijayraghav-io** [comment]
Hi @Amit91848 , it should be working fine now :
https://www.loom.com/share/cda773b1e1604cd19430581d9bed5362?sid=4d346e40-3ab4-4687-b65b-4d295b603d48

File: packages/features/bookings/lib/handleNewBooking/getBookingData.ts [hunk 1]
```diff
@@ -68,7 +68,12 @@ export async function getBookingData<T extends z.ZodType>({
     name: responses.name,
     email: responses.email,
     guests: responses.guests ? responses.guests : [],
-    location: responses.location?.optionValue || responses.location?.value || "",
+    location:
+      typeof responses.location === "object"
+        ? responses.location?.optionValue || responses.location?.value || ""
+        : typeof responses.location === "string"
```
Comment (Amit91848) [inline packages/features/bookings/lib/handleNewBooking/getBookingData.ts:74]: This is still a work around and does not address the main issue. When you change the label, it updates the type of location from radioInput to address which is what throws the error. Please fix that instead.

Screenshot 2024-07-12 125722

**Review by Amit91848 — DISMISSED** [review]
Left a comment.

**Comment by vijayraghav-io** [comment]
Root cause and detailed steps to recreate:

1. Create a new eventType
2. Edit the new eventType created
    1. Edit -> EventSetup (Tab) -> Location - change to ‘In Person (Attendee address)’
    Screenshot 2024-07-12 at 12 57 31 PM

    2. Things will be fine until now, and everything works just fine until you go to the ‘Advanced’ Tab and click on ‘Edit’ Button for Location inside Booking questions
    3. Once you click on Edit button. The following LOC will change type of booking field from radioinput to string
    https://github.com/calcom/cal.com/blob/c7c5cdf133bca0393d00314227de33957fb1e594/packages/features/form-builder/FormBuilder.tsx#L245
   Screenshot 2024-07-12 at 1 09 49 PM

    4. But the booking form expects the location to still be object , Hence the input_error was occurring without much detailed message
    
3. Hence the zod-util.ts file is updated with z.union so that the location field can be either an object or a string. This accomodates the location field to be either object or string  and elimanates input_error in booking input form.
4. Also getBookingData.ts is updated to show the exact address if location is a string instead of radio input

File: packages/features/bookings/lib/handleNewBooking/getBookingData.ts (hunk 1 above)
Comment (vijayraghav-io) [reply packages/features/bookings/lib/handleNewBooking/getBookingData.ts:74]: Yes this is expected to happen as per https://github.com/calcom/cal.com/blob/c7c5cdf133bca0393d00314227de33957fb1e594/packages/features/form-builder/FormBuilder.tsx#L245
Hence the z.union is made for location field similar to name field, Based on user input , the location can be either an object (radioinput) or a string. This was not accomodated in Booking Form

Comment (Amit91848) [reply packages/features/bookings/lib/handleNewBooking/getBookingData.ts:74]: Ah my bad, I missed that

Comment (vijayraghav-io) [reply packages/features/bookings/lib/handleNewBooking/getBookingData.ts:74]: No issues! @Amit91848 👍

File: packages/features/bookings/lib/handleNewBooking/getBookingData.ts [hunk 2]
```diff
@@ -68,7 +68,12 @@ export async function getBookingData<T extends z.ZodType>({
...
+        ? responses.location
+        : "",
```
Comment (Amit91848) [inline packages/features/bookings/lib/handleNewBooking/getBookingData.ts:76]:
```suggestion
    location: responses.location
      ? typeof responses.location === "object"
        ? responses.location.optionValue || responses.location.value
        : responses.location
      : "",
```

NIT: Since location can only be object or string, if not object then we shouldn't have to check if its a string.

**Comment by Amit91848** [comment]
Changes look good @vijayraghav-io  but would also probably need update in the docs as structure of webhooks and maybe also workflow would change for location. Since @CarinaWolli and @joeauyeung approved #13989, would like a second review from them to know if this is how we should attempt to fix this.

Normal response
`      "location": {
        "label": "location",
        "value": {
          "optionValue": "+919992929299",
          "value": "phone"
        },
        "isHidden": false
      },`
      
 Response after changing the label:
 `      "location": {
        "label": "Your Phone Number",
        "value": "+91992929292",
        "isHidden": false
      },`

**Review by hariombalhara — DISMISSED** [review]
Thanks @vijayraghav-io for the PR. But this isn't how it should be fixed, `location` prop format should stay the same.

The bug is introduced by this PR https://github.com/calcom/cal.com/pull/13989. 


The question is why it is being set as a string in the first place.

**Comment by vijayraghav-io** [comment]
> Thanks @vijayraghav-io for the PR. But this isn't how it should be fixed, `location` prop format should stay the same.
> 
> The bug is introduced by this PR #13989.
> 
> The question is why it is being set as a string in the first place.

@hariombalhara , ok this change (changing type to string) was already there , so dint analyze why this was made.
If requirement is not to change the `location` prop format, then there has to be a different solution.
I am looking into older PRs to find why this change was made.

**Comment by vijayraghav-io** [comment]
@hariombalhara  Analysis done:

The fix https://github.com/calcom/cal.com/pull/13989 is not good fix for https://github.com/calcom/cal.com/issues/13670
This fix tries to unhide the location question in booking questions (advanced tab) when `inPersonAttendee` location  is selected (which was originally made hidden by this PR https://github.com/calcom/cal.com/pull/7734).

Since just by unhiding - the other requirements mentioned in https://github.com/calcom/cal.com/issues/13670 , like make the field required.. are not met (in below img).
So whoever fixed might have thought that this requirements can be easily met by converting field type from `radioinput` to `string`, without may be trying (testing) to book an event with this fix. Then, the `input error` (main problem reported in this issue) would have been caught.

Screenshot 2024-07-18 at 1 45 36 PM

Also if user provides a custom label ("Your Address') by editing the location booking question. This was not respected (just by unhiding the bookingField).


To summarize:
1. The requirement of - hiding the location Bookingfield (or Booking question) if there is only one option (mentioned in https://github.com/calcom/cal.com/pull/7734 ) should be revisited. I guess the requirement could be like hide the location field when there is no input required (say for video apps, only link is enough without any input reqd from user).
2. If above implementation is reversed , still some updates have to be done to respect the custom label (like 'Your address') entered in Booking questions (Advanced Tab) to be shown in booking Form.

**Comment by vijayraghav-io** [comment]
@hariombalhara,

Updated with following changes, please check and share your comments -

1. Undone the changes from this PR - https://github.com/calcom/cal.com/pull/13989

2. Removed the usage of `hideWhenJustOneOption` property. Functionally this was used for two purposes -
    i. To remove the `location` booking field itself from Advanced Tab, when `location` had one option. 
        Actually this is not required, but hidden property on `location` booking field had to be set.
    ii. In `getBookingResponsesSchema()-> preprocess() -> schema.superRefine()`.
        This is not effective, as actually the field.hidden is set in `getAndUpdateNormalizedValues()` when there is only one non-input option.
        https://github.com/calcom/cal.com/blob/e6a16f5b5c69a4bc1a8c4b8819183b20c0ef1a21/packages/features/form-builder/FormBuilderField.tsx#L170

3. Updated to show custom label as set in `Advanced Tab` for `location booking field` in Booking Form. This will work regardless of number of location options selected.

4. Updated to set `hidden` and `required` properties of `location booking field` to be shown in `Advanced Tab` , based on the selection and number of location options in `Event Setup` tab. 
Even without this, the rendering of location fields in Booking form works fine. This is to show consistent data while editing the eventType and for the issue https://github.com/calcom/cal.com/issues/13670 , that when only `In person Attendee` location is selected, the `location booking field` should show as required in `Advanced Tab`.

**Comment by vijayraghav-io** [comment]
@hariombalhara , this is w.r.t testing

Tested for following functionalities and is working fine :

1. Location field to be hidden in Booking Form, if there is only one non-input option. (This was already implemented)
2. The custom label entered in `Advanced Tab` (while editing the EventType) for `location Booking field` is respected and rendered in `Booking Form` while booking. This works the same irrespective of number of location options selected.
3. While editing the EventType, if `In Person Attendee` is selected as location in `Event Setup` tab, the `'location booking field` is shown in `Advanced Tab`. 
Also it is set as `required` if this is the only location option.
Will be set as `hidden`, if any conferencing app is the only option.
For all other multiple location options, default to required=false and hidden=false. And will be taken care during the booking form rendering.

**Comment by hariombalhara** [comment]
@vijayraghav-io Thanks for the patience. I have started looking into it now and changed a few things in a single commit. Let me know if it makes sense to you. 

I will be testing it more and probably add a few  tests as well before merging it.

File: apps/web/components/eventtype/Locations.tsx [hunk 3]
```diff
@@ -268,23 +268,6 @@ const Locations: React.FC<LocationsProps> = ({
...
-                        existingBookingFields[findLocation] = {
-                          ...existingBookingFields[findLocation],
-                          type: "radioInput",
-                          label: "",
-                          placeholder: "",
-                        };
-                        setValue("bookingFields", existingBookingFields, {
-                          shouldDirty: true,
-                        });
-                      }
```
Comment (hariombalhara) [inline apps/web/components/eventtype/Locations.tsx:287]: Reverting the code in the faulty PR. No new addition here.

FormBuilder has a dataStore property by which it is made aware of locations and then FormBuilder or the component using the FormBuilder component should do any transformations.

File: packages/features/form-builder/FormBuilder.tsx [hunk 4]
```diff
@@ -45,6 +44,13 @@ function getCurrentFieldType(fieldForm: UseFormReturn<RhfFormField>) {
   return fieldTypesConfigMap[fieldForm.watch("type") || "text"];
 }
 
+function isRequiredField(field: RhfFormField) {
+  // 'location' must be shown as required in the UI(but at backend it is not required because not specifying it defaults to CalVideo)
+  // So, handle it in UI only instead of updating bookingFields config in getBookingFields
+  // TODO: FormBuilder must not be made aware of BookingFields, so move this "location" check to outside FormBuilder where the component is used.
+  return field.name === "location" ? true : field.required;
```
Comment (hariombalhara) [inline packages/features/form-builder/FormBuilder.tsx:51]: This makes a new change as well where "location" field is shown as required now. This created confusion earlier for some users.

File: packages/features/form-builder/FormBuilder.tsx [hunk 5]
```diff
@@ -119,29 +125,37 @@ export const FormBuilder = function FormBuilder({
...
+              // TODO: The dataStore should itself provide the label directly
+              // This is important because FormBuilder isn't aware of what a location is. It is a generic Form Builder
+              const sourceLabel = field.getOptionsAt === "locations" ? "Location" : field.getOptionsAt;
+              options.forEach((option) => {
+                sources.push({
+                  id: option.value,
+                  label: sourceLabel,
+                  type: "system",
+                });
+              });
```
Comment (hariombalhara) [inline packages/features/form-builder/FormBuilder.tsx:142]: image
This shows a badge that this field is coming due to locations. This is similar to what we do on fields added through workflows

File: packages/features/form-builder/FormBuilder.tsx [hunk 6]
```diff
@@ -119,29 +125,37 @@ export const FormBuilder = function FormBuilder({
...
-
-            const { hidden } = getAndUpdateNormalizedValues({ ...field, options }, t);
-            if (field.hideWhenJustOneOption && (hidden || !options?.length)) {
+            const numOptions = options?.length ?? 0;
+            const firstOptionInput =
+              field.optionsInputs?.[options?.[0]?.value as keyof typeof field.optionsInputs];
+            const doesFirstOptionHaveInput = !!firstOptionInput;
+            // If there is only one option and it doesn't have an input required, we don't show the Field for it.
+            // Because booker doesn't see this in UI, there is no point showing it in FormBuilder to configure it.
+            if (field.hideWhenJustOneOption && numOptions <= 1 && !doesFirstOptionHaveInput) {
```
Comment (hariombalhara) [inline packages/features/form-builder/FormBuilder.tsx:154]: We need to still use `hideWhenJustOneOption` because it is used in getBookingResponsesSchema as well.

Also, this makes the code handle any other radioInput kind of custom field(when we allow that in future)

File: packages/features/form-builder/FormBuilderField.tsx [hunk 7]
```diff
@@ -54,6 +54,7 @@ export const FormBuilderField = ({
   const { t } = useLocale();
   const { control, formState } = useFormContext();
 
+  const labelBeforeNormalize = field.label;
```
Comment (hariombalhara) [inline packages/features/form-builder/FormBuilderField.tsx:57]: Why are we using the `labelBeforeNormalize`? What is the difference here?

Comment (vijayraghav-io) [reply packages/features/form-builder/FormBuilderField.tsx:57]: This is required to respect the custom label entered for location in BookingQuestions, because the `getAndUpdateNormalizedValues()` fn will modify the field.label to default label `location`.

**Comment by vijayraghav-io** [comment]
> @vijayraghav-io Thanks for the patience. I have started looking into it now and changed a few things in a single commit. Let me know if it makes sense to you.
> 
> I will be testing it more and probably add a few tests as well before merging it.

Thankyou! @hariombalhara , i am looking through the changes and replied to comment.

File: packages/features/form-builder/FormBuilderField.tsx (hunk 7 above)
Comment (hariombalhara) [reply packages/features/form-builder/FormBuilderField.tsx]: I have fixed the logic so that we use field.label only. The issue was because we were setting field.label to empty string for radioInput type field(e.g. location) when it has just one option and that is an input.

Now, we are not modifying that, and introduced a separate property noLabel and doing some prop drilling.

File: packages/features/form-builder/Components.tsx [hunk 8]
```diff
@@ -398,17 +406,17 @@ export const Components: Record<FieldType, Component> = {
 
       const { t } = useLocale();
 
-      const getCleanLabel = (option: { label: string; value: string }): string | JSX.Element => {
-        if (!option.label) {
+      const getCleanLabel = (label: string): string | JSX.Element => {
```
Comment (hariombalhara) [inline packages/features/form-builder/Components.tsx:409]: We need just the `label`. This way it works even if I have just the label.

File: packages/features/form-builder/Components.tsx [hunk 9]
```diff
@@ -398,17 +408,24 @@ export const Components: Record<FieldType, Component> = {
 
       const { t } = useLocale();
 
-      const getCleanLabel = (option: { label: string; value: string }): string | JSX.Element => {
-        if (!option.label) {
+      const didUserProvideLabel = (
```
Comment (hariombalhara) [inline packages/features/form-builder/Components.tsx:411]: This is who we decide if the organizer has changed the label and if the label has been changed by the organizer, we show the organizer's label instead of the first option label.

File: packages/features/form-builder/FormBuilder.tsx [hunk 10]
```diff
@@ -244,16 +257,7 @@ export const FormBuilder = function FormBuilder({
...
-                        const fieldToEdit = field;
-                        // For radioInput type, when there's only one option, the type and required takes the only first options values
-                        if (fieldToEdit.type === "radioInput" && options.length === 1) {
-                          fieldToEdit.type =
-                            fieldToEdit.optionsInputs?.[options[0].value].type || fieldToEdit.type;
-                          fieldToEdit.required =
-                            fieldToEdit.optionsInputs?.[options[0].value].required || fieldToEdit.required;
-                        }
-
-                        editField(index, fieldToEdit);
```
Comment (hariombalhara) [inline packages/features/form-builder/FormBuilder.tsx:256]: Reverted code from faulty PR.

File: packages/features/form-builder/FormBuilder.tsx [hunk 11]
```diff
@@ -119,29 +125,37 @@ export const FormBuilder = function FormBuilder({
...
+            if (field.hideWhenJustOneOption && numOptions <= 1 && !doesFirstOptionHaveInput) {
               return null;
             }
-            let fieldType = fieldTypesConfigMap[field.type];
-            let isRequired = field.required;
-            // For radioInput type, when there's only one option, the type and required takes the first options values
-            if (field.type === "radioInput" && options.length === 1) {
-              fieldType = fieldTypesConfigMap[field.optionsInputs?.[options[0].value].type || field.type];
-              isRequired = field.optionsInputs?.[options[0].value].required || field.required;
-            }
```
Comment (hariombalhara) [inline packages/features/form-builder/FormBuilder.tsx:144]: Reverted code from faulty PR.

File: packages/features/form-builder/FormBuilderField.tsx [hunk 12]
```diff
@@ -227,7 +244,7 @@ export const ComponentForField = ({
   }
   if (componentConfig.propsType === "text") {
     return (
-      <WithLabel field={field} readOnly={readOnly}>
+      <WithLabel field={field} readOnly={readOnly} noLabel={noLabel}>
```
Comment (hariombalhara) [inline packages/features/form-builder/FormBuilderField.tsx:247]: Pass on `noLabel` directly instead of setting field.label to empty string because that would mean that someone wanting to use the field.label still won't have it.

**Review by CarinaWolli — COMMENTED** [review]
Tested it and everything worked great 👍

**Review by hariombalhara — APPROVED** [review]
